File: 2. Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *head, *p = NULL;
        ListNode *p1 = l1, *p2 = l2;
        int add = 0;
        while (p1 || p2 || add) {
            int t = 0;
            if (p1)
                t += p1->val;
            if (p2)
                t += p2->val;
            t += add;
            add = t/10;
 
            if (p != NULL) {
                p->next = new ListNode(t%10);
                p = p->next;
            } else {
                p = new ListNode(t%10);
                head = p;
            }

            if (p1) p1 = p1->next;
            if (p2) p2 = p2->next;
        }
        
        return head;
    }
};

File: go/src/leetcode/2. Add Two Numbers/2. Add Two Numbers.go
package main

type ListNode struct {
	Val int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{0, nil}
	p := dummy

	r, val := 0, 0
	p1, p2 := l1, l2
	for ; p1 != nil || p2 != nil; {
		if p1 != nil && p2 != nil {
			val = p1.Val + p2.Val + r
			p1.Val = val % 10
			r = val / 10
			p.Next = p1
			p1 = p1.Next
			p2 = p2.Next
		} else if p1 != nil {
			val = p1.Val + r
			p1.Val = val % 10
			r = val / 10
			p.Next = p1
			p1 = p1.Next
		} else {
			val = p2.Val + r
			p2.Val = val % 10
			r = val / 10
			p.Next = p2
			p2 = p2.Next
		}
		p = p.Next
	}

	if r != 0 {
		p.Next = &ListNode{r, nil}
	}

	return dummy.Next
}
